src/main/java/org/cryptacular/util/CipherUtil.java

```diff
@@ -240,13 +242,62 @@ public static void decrypt(
     final OutputStream output)
     throws CryptoException, EncodingException, StreamException
   {
-    final CiphertextHeader header = CiphertextHeader.decode(input);
+    final CiphertextHeader header = decodeHeader(input, String -> key);
     final PaddedBufferedBlockCipher padded = new PaddedBufferedBlockCipher(cipher, new PKCS7Padding());
     padded.init(false, new ParametersWithIV(new KeyParameter(key.getEncoded()), header.getNonce()));
     process(new BufferedBlockCipherAdapter(padded), input, output);
   }
 
 
+  /**
+   * Decodes the ciphertext header at the start of the given byte array.
+   * Supports both original (deprecated) and v2 formats.
+   *
+   * @param  data  Ciphertext data with prepended header.
+   * @param  keyLookup  Decryption key lookup function.
+   *
+   * @return  Ciphertext header instance.
+   */
+  public static CiphertextHeader decodeHeader(final byte[] data, final Function<String, SecretKey> keyLookup)
+  {
+    try {
+      return CiphertextHeaderV2.decode(data, keyLookup);
+    } catch (EncodingException e) {
+      return CiphertextHeader.decode(data);
+    }
```

Comment:
I feel like CiphtertextHeaders should expose a `isInstance` method or some such thing. This will set you up for a cleaner implementation if a v3 ever arises. Not a fan of the try-catch here.

Reply:
Factory pattern is probably best. I'll take a quick stab at it.

Reply:
Upon review and further consideration, I think this is the best solution at present when taking into account the stream case. Given that not all streams are buffered, attempting to read the first 4 bytes of the message to determine the version could break the subsequent decode methods on concrete classes. In next minor (or major) version I think we'll just make `CiphertextHeader` an interface and give it the full treatment with a factory class that handles decoding.